src/main/java/org/jenkinsci/plugins/mesos/MesosSlave.java

```diff
@@ -27,32 +29,58 @@
   // Holds the current USI status for this agent.
   Optional<PodStatus> currentStatus = Optional.empty();
 
+  private final Boolean reusable;
+
+  private final MesosCloud cloud;
+
+  private final ExecutorService executorService;
+
+  private final String podId;
+
   public MesosSlave(
-      String name,
+      MesosCloud cloud,
+      String id,
       String nodeDescription,
       String labelString,
       List<? extends NodeProperty<?>> nodeProperties)
       throws Descriptor.FormException, IOException {
     super(
-        name,
+        id,
         nodeDescription,
-        null,
+        "jenkins",
         1,
-        null,
+        Mode.NORMAL,
         labelString,
         new JNLPLauncher(),
         null,
         nodeProperties);
+
+    // pass around the MesosApi connection via MesosCloud
+    this.cloud = cloud;
+    this.reusable = true;
+    this.executorService = Executors.newSingleThreadExecutor();
+    this.podId = id;
   }
 
   /**
    * Polls the agent until it is online. Note: This is a non-blocking call in contrast to the
    * blocking {@link AbstractCloudComputer#waitUntilOnline}.
    *
-   * @return This agent when it's online.
+   * @return The future agent that will come online.
    */
-  public CompletableFuture<MesosSlave> waitUntilOnlineAsync() {
-    throw new NotImplementedException();
+  public CompletableFuture<Node> waitUntilOnlineAsync() {
+    return CompletableFuture.supplyAsync(
+        () -> {
+          await()
+              .with()
+              .pollInterval(2, TimeUnit.SECONDS)
+              .and()
+              .pollDelay(1, TimeUnit.SECONDS)
+              .atMost(5, TimeUnit.MINUTES)
+              .until(this::isRunning);
```

Comment:
`this::isRunning` is not correct. We have to wait until the agent `isOnline`. Also, please use `Source.tick` as @meln1k mentioned.